### PR TITLE
added karatsuba multiplication in python along with readme

### LINKED ISSUE #1581: Adding python implementation of karatsuba multiplication along with readme
Karatsuba multiplication is a technique to multiply large numbers efficiently(it does simple multiplication too, ofcourse). I think we could implement it in python and add an explanation along with it(may be in the form of a readme file), so that it would be helpful for everyone.

### PR DESCRIPTION
**Pull request by lakshyajit165**
Fixes
Implemented karatsuba multiplication in python along with readme

Thanks!

### CODE REVIEW

File: Karatsuba_Algo/karatsuba.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/python3
+
+from math import ceil, floor
+
+
```
Comment (SakshiShreya) [inline Karatsuba_Algo/karatsuba.py]: Remove the extra new-line.

File: Karatsuba_Algo/karatsuba.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+def karatsuba(x,y):
+    #base case
```
Comment (SakshiShreya) [inline Karatsuba_Algo/karatsuba.py]: Add a space after # in all the comments.

File: Karatsuba_Algo/karatsuba.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    if x < 10 and y < 10: # in other words, if x and y are single digits
+        return x*y
+
+    n = max(len(str(x)), len(str(y)))
+    m = ceil(n/2)   #Cast n into a float because n might lie outside the representable range of integers.
+
+    a = floor(x / 10**m)
```
Comment (SakshiShreya) [inline Karatsuba_Algo/karatsuba.py]: Operators should have space around them. For more clarity, you can use brackets.

File: Karatsuba_Algo/karatsuba.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    #recursive steps
+    ac = karatsuba(a,c)
+    bd = karatsuba(b,d)
+    ad_plus_bc = karatsuba(a + b, c + d) - ac - bd
+
+    return int(ac*(10**(m*2)) + ad_plus_bc*(10**m) + bd)
+
+
+#a,b = map(int, input().split(' '))
```
Comment (SakshiShreya) [inline Karatsuba_Algo/karatsuba.py]: Please mention INPUT and OUTPUT instead of writing the whole sentence.

**Comment by lakshyajit165** [comment]
I have made changes according to your review. Please check

File: Karatsuba_Algo/explanation.md [hunk 5]
```diff
@@ -0,0 +1,20 @@
+# Karatsuba Multiplication
+<p>Karatsuba multiplication is an effective and efficient way of multiplication, which comes handy when multiplying really large numbers. It is different from the steps followed in conventional multiplication technique due to the fact that, it recursively performs the multiplication, thereby, making it possible to accomodate really large numbers.</p>
```
Comment (jainaman224) [inline Karatsuba_Algo/explanation.md]: Don't use html tags in markdown.

File: Karatsuba_Algo/explanation.md [hunk 6]
```diff
@@ -0,0 +1,20 @@
+# Karatsuba Multiplication
```
Comment (jainaman224) [inline Karatsuba_Algo/explanation.md]: Change filename to Readme.md. It is standard name used mostly everywhere.

File: Karatsuba_Algo/explanation.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+- Consider the above 2 numbers, let x = 5678 and y = 1234
+- Break each number into 2 halves and store them in variables
+- In this case, <strong>let a = 56, b = 78, c = 12, d = 34</strong>
+- Now we can express x and y as, <strong>x = 10<sup>(n/2)</sup> * a + b</strong> and <strong>y = 10<sup>(n/2)</sup> * c + d</strong>, where a,b,c,d are n/2 digit numbers.
+- Therefore, <strong>x * y = 10<sup>n</sup>*a*c + 10<sup>(n/2)</sup> * (a*d + b*c) + b*d</strong>
+- Now, on the first look it may seem that there are 4 recursive calls, to compute ac, ad, bc and bd - but here is the trick:
+- It's also called as Gauss's trick - which requires only 3 recursive calls.
+- If we carefully observe, we have already computed ac and bd in the first 2 steps, so if we subtract this result from the result of <strong>(a+b) * (c+d)</strong>, we reduce 1 recursive call.
+- It may seems strange, but when the computations to be performed are of very high level(multiplying really large digit numbers in this case), reducing even a single recursive call significantly affects performance.
+- Checkout the code in python, to understand the recursive work flow better.
```
Comment (jainaman224) [inline Karatsuba_Algo/explanation.md]: Add a code section.

File: Karatsuba_Algo/karatsuba.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/python3
```
Comment (jainaman224) [inline Karatsuba_Algo/karatsuba.py]: Fix filename.

File: Karatsuba_Algo/karatsuba.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    d = y % (10**m)
+
+    # recursive steps
+    ac = karatsuba(a,c)
+    bd = karatsuba(b,d)
+    ad_plus_bc = karatsuba(a + b, c + d) - ac - bd
+
+    return int(ac * (10**(m * 2)) + ad_plus_bc * (10**m) + bd)
+
+print(karatsuba(123456789, 987654321)) # INPUT
```
Comment (jainaman224) [inline Karatsuba_Algo/karatsuba.py]: Take input dynamically.

**Comment by lakshyajit165** [comment]
@SakshiShreya and @jainaman224. I have made the changes as required. Kindly review them.

File: Karatsuba_Algo/karatsuba.py (hunk 8 above)
Comment (jainaman224) [reply Karatsuba_Algo/karatsuba.py]: File and folder name should be Karatsuba_Algorithm

File: Karatsuba_Algo/karatsuba.py (hunk 9 above)
Comment (jainaman224) [reply Karatsuba_Algo/karatsuba.py]: Not fixed. Input should be taken from user.

**Comment by lakshyajit165** [comment]
@SakshiShreya  and @jainaman224. I have changed the folder name as suggested. Please review.

File: Karatsuba_Algo/explanation.md (hunk 6 above)
Comment (jainaman224) [reply Karatsuba_Algo/explanation.md]: Not done.

File: Karatsuba_Algo/karatsuba.py (hunk 8 above)
Comment (jainaman224) [reply Karatsuba_Algo/karatsuba.py]: Not done.

File: Karatsuba_Algo/explanation.md (hunk 7 above)
Comment (jainaman224) [reply Karatsuba_Algo/explanation.md]: Not done.
It meant, there should be a implementation section.

Comment (lakshyajit165) [reply Karatsuba_Algo/explanation.md]: @jainaman224. As discussed with @asha15, any algorithm should contain the following:
Explanation
Algorithm
Pseudocode
Example
Time Complexity
I saw that the Time Complexity part of karatsuba algorithm was missing. I have added that now. The rest of the above mentioned, are present as of my knowledge. Please review.

Comment (lakshyajit165) [reply Karatsuba_Algo/explanation.md]: @jainaman224. The steps of the algorithm is supposedly the implementation section. Do I need to name it that way?

File: Karatsuba_Algo/explanation.md (hunk 6 above)
Comment (lakshyajit165) [reply Karatsuba_Algo/explanation.md]: I have changed the filename. @asha15 and @jainaman224, Please review.

File: Karatsuba_Algorithm/Readme.md [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    return int(ac*(10**(m*2)) + ad_plus_bc*(10**m) + bd)
+```
+
+### Complexity
+
+To analyze the complexity of the Karatsuba algorithm, consider the number of multiplications the algorithm performs as a function of n, M(n).
+
+The algorithm multiplies together two nn-bit numbers. If n=2^k for some k then the algorithm recurses three times on n/2-bit number. The recurrence for this is:
+
+M(n) = 3M(n/2)
```
Comment (asha15) [inline Karatsuba_Algorithm/Readme.md]: keep space around operators

File: Karatsuba_Algorithm/Readme.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+This takes care of the multiplications required for Karatsuba--now to consider the additions and subtractions. There are O(n)O(n) additions and subtractions required for the algorithm. Therefore, the overall recurrence for the Karatsuba algorithm is:
+
+T(n) = 3T(n/2) + O(n)
```
Comment (asha15) [inline Karatsuba_Algorithm/Readme.md]: keep space around operator

File: Karatsuba_Algorithm/Readme.md [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+Using the master theorem on the above recurrence yields that the running time of the Karatsuba algorithm is: $\Theta$(n<sup>log<sub>2</sub>3</sup>)
```
Comment (asha15) [inline Karatsuba_Algorithm/Readme.md]: `$\Theta$(n<sup>log<sub>2</sub>3</sup>)`
the algorithm is not clear.

File: Karatsuba_Algorithm/karatsuba.py [hunk 13]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/python3
+from math import ceil, floor
+
+def karatsuba(x,y):
```
Comment (asha15) [inline Karatsuba_Algorithm/karatsuba.py]: keep space after comma

File: Karatsuba_Algorithm/karatsuba.py [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    # base case
+    if x < 10 and y < 10: # in other words, if x and y are single digits
+        return x*y
```
Comment (asha15) [inline Karatsuba_Algorithm/karatsuba.py]: keep space around operator

File: Karatsuba_Algorithm/karatsuba.py [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+
+    n = max(len(str(x)), len(str(y)))
+    m = ceil(n/2)   # Cast n into a float because n might lie outside the representable range of integers.
```
Comment (asha15) [inline Karatsuba_Algorithm/karatsuba.py]: keep space around operator

File: Karatsuba_Algorithm/karatsuba.py [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+    c = floor(y / 10**m)
+    d = y % (10**m)
+
+    # recursive steps
+    ac = karatsuba(a,c)
+    bd = karatsuba(b,d)
+    ad_plus_bc = karatsuba(a + b, c + d) - ac - bd
+
+    return int(ac*(10**(m*2)) + ad_plus_bc*(10**m) + bd)
```
Comment (asha15) [inline Karatsuba_Algorithm/karatsuba.py]: keep space around operator

File: Karatsuba_Algorithm/karatsuba.py [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+
+# INPUT 123456789, 987654321
+a,b = map(int, input().split(' '))
```
Comment (asha15) [inline Karatsuba_Algorithm/karatsuba.py]: keep space after comma

File: Karatsuba_Algorithm/Readme.md (hunk 12 above)
Comment (lakshyajit165) [reply Karatsuba_Algorithm/Readme.md]: Do I need to explain the derivation from master method here too?

Comment (asha15) [reply Karatsuba_Algorithm/Readme.md]: no,i asked you to replace this
` Θ ( n log ⁡ 2 3 ) ≈ Θ ( n 1.585 ) `

**Pull request closed without merging**